filebeat/reader/line/decoder.go

```diff
@@ -0,0 +1,166 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package line
+
+import (
+	"fmt"
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []uint8
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) *decoderReader {
+	return &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		bufferSize: bufferSize,
+	}
+}
+
+func (r *decoderReader) read(buf []byte) (int, error) {
+	b := make([]byte, r.bufferSize)
+
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
```

Comment:
nit: what if `buf.Len()` == 1? In this case we would copy only one byte and return. It's not incorrect, but a little inefficient. Let's add a `TODO` comment at least, noting that we should try to collect at least `len(buf) - r.buf.Len()` bytes, if `len(buf) > r.buf.Len()`.